ducktape/tests/runner_client.py

```diff
@@ -0,0 +1,281 @@
+# Copyright 2016 Confluent Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import logging
+import os
+import time
+import traceback
+import zmq
+
+from ducktape.tests.message import Request
+from ducktape.tests.loader import TestLoader
+from ducktape.tests.serde import SerDe
+from ducktape.tests.test import TestLogger
+
+from ducktape.tests.result import TestResult, IGNORE, PASS, FAIL
+from ducktape.tests.reporter import SingleResultFileReporter
+from ducktape.utils.local_filesystem_utils import mkdir_p
+
+
+def run_client(port, logger_name, log_dir, debug, max_parallel):
+    client = RunnerClient(port, logger_name, log_dir, debug, max_parallel)
+    client.run()
+
+
+class RunnerClient(object):
+    """Run a single test"""
+
+    def __init__(self, port, logger_name, log_dir, debug, max_parallel):
+        self.serde = SerDe()
+        self.logger = TestLogger(logger_name, log_dir, debug, max_parallel).logger
+        self.runner_port = port
+
+        self.id = "test-runner-%d-%d" % (os.getpid(), id(self))
+        self.request = Request(self.id)
+        self.sender = Sender("localhost", str(self.runner_port), self.logger)
+
+        ready_reply = self.sender.send(self.request.ready())
+        self.session_context = ready_reply["session_context"]
+        self.test_metadata = ready_reply["test_metadata"]
+        self.cluster = ready_reply["cluster"]
+
+        # Wait to instantiate the test object until running the test
+        self.test = None
+        self.test_context = None
+
+    def send(self, event):
+        return self.sender.send(event)
+
+    def collect_test_context(self, directory, file_name, cls_name, method_name, injected_args):
+        # TODO - different logger for TestLoader object
+        loader = TestLoader(self.session_context, self.logger, injected_args=injected_args, cluster=self.cluster)
+        loaded_context_list = loader._discover(directory, file_name, cls_name, method_name)
+
+        assert len(loaded_context_list) == 1
+        test_context = loaded_context_list[0]
+        test_context.cluster = self.cluster
+        return test_context
+
+    def run(self):
+        self.log(logging.INFO, "Loading test %s" % str(self.test_metadata))
+        self.test_context = self.collect_test_context(**self.test_metadata)
+
+        self.send(self.request.running())
+        if len(self.cluster) != self.cluster.num_available_nodes():
```

Comment:
Strictly speaking I don't think this is always true. `LocalhostCluster` initializes the available nodes to `sys.maxint` and also always returns `sys.maxint` for `len`. JsonCluster uses its explicit list to keep everything in sync.

We may just not have clearly defined the semantics, but I don't think this test will always work. A `DockerCluster` might be another case similar to `LocalhostCluster` where there isn't a fixed size up front, and it could effectively have an infinitely large cluster.